sim/battle.ts

```diff
@@ -3075,20 +3087,20 @@ export class Battle extends Dex.ModdedDex {
 		return team as PokemonSet[];
 	}
 
-	setPlayer(slot: 'p1' | 'p2', options: PlayerOptions) {
+	setPlayer(slot: SideID, options: PlayerOptions) {
 		let side;
 		let didSomething = true;
-		if (!this[slot]) {
+		const slotNum = parseInt(slot[1], 10) - 1;
+		if (!this.sides[slotNum]) {
 			// create player
-			const slotNum = (slot === 'p2' ? 1 : 0);
 			const team = this.getTeam(options.team || null);
 			side = new Side(options.name || `Player ${slotNum + 1}`, this, slotNum, team);
 			if (options.avatar) side.avatar = '' + options.avatar;
-			this[slot] = side;
+			if (slot === 'p1' || slot === 'p2') this[slot] = side;
```

Comment:
Let's deprecate+remove the existing `.p1` and `.p2` properties.

Just remember to update `>>>` (eval).

Reply:
Oh, ok.

Reply:
Yeah, so practically every battle test relies on this API, and `grep sed` tricks don't work on Windows 🙃

Reply:
Never mind, I installed `gow` so I was able to change all the tests with one command! I hope that's ok...

Reply:
LOL

I actually might keep them around for tests, actually. Or, like, have the tests manually add `.p1` and `.p2` functions for usability. I'd consider that a better approach than rewriting a fuckton of tests.

Reply:
Having the tests manually add them would still require rewriting a fuckton of tests, no? Or at least, rewriting a fuckton of files. That sounds like a lot more work than what I did.

Reply:
No, it'd just involve editing `test/common.js`

Reply:
That doesn't work because there are still a bunch of tests that use the deprecated `Battle#join` API, and by having the teams after the battle is created, there's no way for anything in `test/common.js` to have a reference to those `Side` objects.

Reply:
So either way it's a bunch of tests edited.

Reply:
I muddled my changes together by mistake, and now everything hinges on how this is handled, so I need you to make a call on what to do with the tests before I can push anything. One option is to just leave those old tests as is for now, and make them store the return value of `Battle#join`, which is the very side object they're looking for.

Reply:
I've decided to just go ahead and manually edit all the tests that still relied on both the `Battle#p1` and the `Battle#join` APIs, which is 570 tests. That should give you plenty of time to finish reviewing #5216 😉

Reply:
Is `Battle#join` actually deprecated? I didn't even know.

Reply:
You deprecated it https://github.com/Zarel/Pokemon-Showdown/blame/f3e45fbb72ddcccbda267428cdb7d6430a31b6d2/sim/battle.js#L3408

Reply:
Oh yeah, the new API is `setPlayer`.

Reply:
btw can you edit the tests in a separate PR? Smaller PRs are generally easier to refactor.

Reply:
Not easily. Like I said, I muddled my changes together and splitting them apart at this point would be like unbaking a cake.

Reply:
Never mind, it turned out to be more like unmaking a sandwich.